charts/linkerd2/values.yaml

```diff
@@ -166,7 +166,11 @@ heartbeatSchedule: "0 0 * * *"
 
 # proxy injector configuration
 proxyInjector:
+  enableNamespacesByDefault: false
```

Comment:
Wouldn't this disable the injector for everyone by default?

Reply:
Sort of, the linkerd proxy injector mutating webhook was called for all pod creation events. By default, it would then filter to only those pods with the linkerd.io/inject:true label k/v pair. 

This was the best option before k8s 1.15. However, if the linkerd proxy service was down for any reasons, you are unable to create pods. Provisioning tools like kops create pods for the control plane. It created a chicken and egg problem where linkerd blocked the creation of kube-apiserver pod and linkerd could not receive requests b/c of a partially functioning kube-apiserver. The same situation happens if you use cert-manager to rotate linkerd certificates.

Reply:
The chicken/egg problem is the reason we recommend skipping `kube-system` in the HA docs (as by default this isn't actually an issue).

Reply:
Those docs can be removed by ensuring a default install of linkerd does not break pod creation for the cluster.

Reply:
Do you mean the `--ha` install? The default install of Linkerd shouldn't break pod creation on the cluster.

Reply:
Does `--ha` change the mutatingwebhook? The issue we faced is that all pod events go to proxy injector unless specific namespace labels are included. Having the proxy injector ignore specific pods is not enough to ensure pod creations work.

Reply:
Yes, the default MWC policy is "ignore" - if the proxy injector is unreachable for any reason, the API server will just skip it and create the pods.

Reply:
That would workaround the issue but it also would lead to inconsistent application of linkerd proxies on pods which is not ideal. We want to ensure that pods that should have linkerd get them and those pods not participating (like kube-system) are not impacted by linkerd.

If the issue is with changing the default behavior, we certainly can look at just the objectSelector. I found that Istio had taken an optin approach which ensures minimal cluster impact.

Reply:
From my 3 previous questions, we need to answer:

- How do we preserve the current behavior of it being on by default?
- What happens when we use this on a k8s 1.14 cluster?
- Does it work? The docs suggest that this is label based, and we're annotation based.

I think it is a great addition to the `--ha` configuration as it is always tough to remember to opt-out `kube-system`. We just can't change the default install behavior.

Reply:
- The existing default install of linkerd changes the cluster and is risky for anyone to use.
- Helm has k8s version flags, which is in the PR
- Object selectors do not work on annotations. Annotations are not intended to be used for selecting objects, they are for annotating the object with additional metadata. https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#attaching-metadata-to-objects So we can not provide per pod selection in linkerd as k8s intends it without also moving the inject selection criteria to a label from annotation.